go/libraries/doltcore/merge/merge_prolly_rows.go

```diff
@@ -740,11 +868,113 @@ func newValueMerger(merged, leftSch, rightSch, baseSch schema.Schema, syncPool p
+// matching tag is not found, then this function falls back to looking for a matching column by name. If no
+// matching column is found, then this function returns -1.
+func findNonPKColumnMappingByTagOrName(sch schema.Schema, col schema.Column) int {
+	if idx, ok := sch.GetNonPKCols().TagToIdx[col.Tag]; ok {
+		return idx
+	} else {
+		return findNonPKColumnMappingByName(sch, col.Name)
+	}
+}
+
+// migrateDataToMergedSchema migrates the data from the left side of the merge of a table to the merged schema. This
+// currently only includes updating the primary index. This is necessary when a schema change is
+// being applied, so that when the new schema is used to pull out data from the table, it will be in the right order.
+func migrateDataToMergedSchema(ctx context.Context, tm *TableMerger, vm *valueMerger, mergedSch schema.Schema) error {
+	lr, err := tm.leftTbl.GetRowData(ctx)
+	if err != nil {
+		return err
+	}
+	leftRows := durable.ProllyMapFromIndex(lr)
+	mut := leftRows.Mutate()
+	mapIter, err := mut.IterAll(ctx)
+	if err != nil {
+		return err
+	}
+
+	leftSch, err := tm.leftTbl.GetSchema(ctx)
+	if err != nil {
+		return err
+	}
+	valueDescriptor := leftSch.GetValueDescriptor()
+
+	for {
+		key, value, err := mapIter.Next(ctx)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		pool := vm.syncPool
+		modifiedValue := remapTuple(value, valueDescriptor, vm.leftMapping)
+		modifiedValueAsTuple := val.NewTuple(pool, modifiedValue...)
+		err = mut.Put(ctx, key, modifiedValueAsTuple)
+		if err != nil {
+			return err
+		}
+	}
+
+	m, err := mut.Map(ctx)
+	if err != nil {
+		return err
+	}
+
+	newIndex := durable.IndexFromProllyMap(m)
+	newTable, err := tm.leftTbl.UpdateRows(ctx, newIndex)
+	if err != nil {
+		return err
+	}
+	tm.leftTbl = newTable
+
+	// TODO: for now... we don't actually need to migrate any of the data held in secondary indexes (yet).
+	//       We're currently dealing with column adds/drops/renames/reorders, but none of those directly affect
+	//       secondary indexes. Columns drops *should*, but currently Dolt just drops any index referencing the
+	//       dropped column, so there's nothing to do currently.
+	//       https://github.com/dolthub/dolt/issues/5641
+	//       Once we start handling type changes changes or primary key changes, or fix the bug above,
+	//       then we will need to start migrating secondary index data, too.
+
+	return nil
+}
+
 // tryMerge performs a cell-wise merge given left, right, and base cell value
 // tuples. It returns the merged cell value tuple and a bool indicating if a
 // conflict occurred. tryMerge should only be called if left and right produce
 // non-identical diffs against base.
 func (m *valueMerger) tryMerge(left, right, base val.Tuple) (val.Tuple, bool) {
+	// We can't merge two divergent keyless rows
 	if m.keyless {
 		return nil, false
 	}
```

Comment:
This is... not great. The cell-wise merge callback in `ThreeWayDiffer` is called when equal keys with unequal values are encountered. For keyless tables, this only occurs when the "cardinality" field of the value tuple has been changed on both sides. For example:
```
key:      val.Tuple(xxHash3({c1, c2, c3}))
ancestor: val.Tuple({cardinality=6, c1, c2, c3})
left:     val.Tuple({cardinality=4, c1, c2, c3})
right:    val.Tuple({cardinality=10, c1, c2, c3})
```
In this case, `left` deleted two copies of the keyless row and `right` added four copies. So the correct merged value tuple should be `val.Tuple({cardinality=8, c1, c2, c3})`

Reply:
ooh, good catch. I just added a new test case to trigger this and pushed a fix in a615e3aeef1f3587db10c5d10b119184d0e25640. Thanks for finding this one!

Reply:
Actually... I was playing with this one some more, and I think returning `nil, false` from `tryMerge` is actually correct here... 

In the case you described, it seems like this should still be a conflict, since we can't decide whether we should delete two rows, add four rows, or something in between, so it needs user involvement to resolve. I tested this with an LD format repo and a DOLT format repo and got that behavior. The [TestKeylessMergeConflicts](https://github.com/dolthub/dolt/blob/4998109a960b541f6037f39772da72c18f202311/go/libraries/doltcore/merge/keyless_integration_test.go#L133) also seem to encode this behavior. 

By returning `nil, false` from the `tryMerge` callback function (which should probably be renamed to `tryResolveMergeConflict` actually...) is called when the conflict is detected, so by not resolving it there, it gets tracked correctly as a conflict for a user to resolve manually. 

Let me know if I just completely misunderstood or misinterpreted the feedback.